### ISSUE TITLE
War plugin's providedRuntime configuration does not inherit runtime dependencies defined in a java-platform bom on Gradle 7

### ISSUE THREAD

**Issue by rspido** [gradle/gradle#17870]
Until Gradle 6, it was possible to declare version-less dependencies in configuration providedRuntime for dependencies that are declared as runtime dependencies in a platform bom.
It no longer works on Gradle 7.1.1

### Expected Behavior
Dependency versions to be resolved as before

### Current Behavior
Gradle 7.1.1 cannot resolve the versions. It fails with the following error:

```
Execution failed for task ':my-web-component:compileJava'.
> Could not resolve all files for configuration ':my-web-component:compileClasspath'.
   > Could not find org.ow2.asm:asm:.
     Required by:
         project :my-web-component
```

### Steps to Reproduce 
Here are two very simple excerpts of a java-platform (bom) module that defines a simple runtime dependency and a web module that consumes it:

_my-platform-bom/build.gradle_
```
plugins {
    id 'java-platform'
}

dependencies {
    constraints {
        runtime 'org.ow2.asm:asm:6.2.1'
    }
}
```

_my-web-component/build.gradle_
```
...
apply plugin: 'war'

dependencies {
    ...
    providedRuntime platform(project(":my-platform-bom"))
    providedRuntime 'org.ow2.asm:asm'
	...
}
...
```

### Workaround
Explicitly declare the versions in the consumer project or, less ideally, define the constraints in the bom with the 'api' scope

**Comment by rspido** [comment]
The root cause of this issue might be https://github.com/gradle/gradle/issues/17415

**Comment by ljacomet** [comment]
The root cause is indeed the same, closing as duplicate of #17415 which will be fixed for Gradle 7.3
